scripts/Run-Self-Hosted-Runner-Test.ps1

```diff
@@ -0,0 +1,261 @@
+    $ProcDumpArguments = "-r -ma $($TestProcess.Id) $UserModeDumpFilePath"
+    Write-Log "Dump Command: $ProcDumpBinaryPath $ProcDumpArguments"
+    $ProcDumpProcess = Start-Process -NoNewWindow `
+        -FilePath $ProcDumpBinaryPath `
+        -ArgumentList $ProcDumpArguments `
+        -Wait -PassThru
+    Write-Log "Waiting for user mode dump to complete..."
+    $ProcDumpProcess.WaitForExit()
+
+    # Get procdump64.exe's exit code.
+    $ExitCode = $($ProcDumpProcess.ExitCode)
+    Write-Log "$ProcDumpBinaryPath completed with exit code: $ExitCode`n"
+
+    # Flush disk buffers to ensure user mode dump data is completely written to disk.
+    Write-Log "User mode dump completed. Flushing disk buffers..."
+    Write-VolumeCache -DriveLetter C
+
+    # Wait for a bit for things to settle down to ensure the user mode dump is completely flushed.
+    Start-Sleep -seconds 10
+
+    # Paranoia check: Make sure a (seemingly) valid dump file _is_ created.
+    $UserModeDumpSizeMB =
+        (((Get-ItemProperty -Path $UserModeDumpFilePath).Length) /1MB).ToString("F2")
+    if ($UserModeDumpSizeMB -eq 0) {
+        Write-Log "* WARNING * User mode dump $UserModeDumpFilePath NOT CREATED"
+    } else {
+        Write-Log "`n`n"
+        Write-Log "Created $UserModeDumpFilePath, size: $UserModeDumpSizeMB MB"
+        Write-Log "`n`n"
+    }
+
+    if ($NeedKernelDump) {
+        Write-Log "Creating kernel dump...`n"
+        # Wait a bit for the above message to show up in the log.
+        Start-Sleep -seconds 5
+
+        # This will/should not return (test system will/should bluescreen and reboot).
+        Start-Process -NoNewWindow -Wait -FilePath $NotMyFaultBinaryPath -ArgumentList "/crash"
+
+        # If we get here, notmyfault64.exe failed for some reason. Kill the hung process, throw error.
+        $TestProcess.Kill()
+        ThrowWithErrorMessage `
+            -ErrorMessage "*** ERROR *** $($PSCommandPath): kernel mode dump creation FAILED"
+    } else {
+        Write-Log "`n`n"
+        Write-Log "Kernel dump not needed, killing test process $($TestProcess.ProcessName)..."
+        $TestProcess.Kill()
+        Write-Log "`n`n"
+
+        $ErrorMessage = "*** ERROR *** $($PSCommandPath): $FullTestCommandSpec.exe exceeded test hang timout of " +
+            "$TestHangTimeout seconds"
+        ThrowWithErrorMessage -ErrorMessage $ErrorMessage
+    }
+} else {
+
+    # Ensure the process has completely exited.
+    $TestProcess.WaitForExit()
+
+    # Read and display the output (if any) from the temporary output file.
+    $TestStandardOutput = Get-Content -Path $TestTempOutFile
+    if ($TestStandardOutput) {
+        Write-Log "`n`n"
+        Write-Log "Test Program output:`n"
+        foreach($Line in $TestStandardOutput) {
+            if ($Line -ne $Null) {
+                Write-Log $Line
+            }
+        }
+    }
+    Write-Log "`n"
+
+    if ($($TestProcess.ExitCode) -ne 0) {
```

Comment:
Looks like this does the failure detection now, but to make sure we don't have false positive in the CI, please include screenshots of succesfuly detected test failures.

Reply:
PR description updated with output from a test failure.